earth_enterprise/src/portableserver/build.py

```diff
@@ -0,0 +1,256 @@
+        else:
+            shutil.copy2(source_path, destination_path)
+
+class Builder(object):
+    """Builds Portable server."""
+
+    def __init__(self, build_dir, source_dir, platform):
+        self.build_dir = build_dir
+        self.source_dir = source_dir
+        self.platform = platform
+
+        self.base_version = None
+        self.build_date = None
+        self.version_string = None
+        self.get_version()
+
+        self.package_dir = os.path.join(
+            self.build_dir,
+            'portableserver-{}-{}'.format(self.platform, self.version_string))
+        self.resources_dir = os.path.join(
+            self.source_dir, 'portableserver', 'resources')
+        self.server_dir = os.path.join(self.package_dir, 'server')
+        self.tar_package_name = \
+            'portableserver-{}-{}'.format(self.platform, self.version_string)
+        self.zip_package_name = self.tar_package_name
+        if self.platform == 'windows':
+            self.should_create_tar_package = False
+            self.should_create_zip_package = True
+        else:
+            self.should_create_tar_package = True
+            self.should_create_zip_package = False
+
+    def build(self):
+        """Builds and packages Portable server."""
+
+        shutil.rmtree(self.build_dir, ignore_errors=True)
+        ensure_directory(self.build_dir)
+        ensure_directory(self.package_dir)
+        ensure_directory(self.server_dir)
+        self.obtain_server_resources()
+        self.build_fileunpacker()
+        self.obtain_sample_globes()
+        if self.should_create_tar_package:
+            self.create_tar_package()
+        if self.should_create_zip_package:
+            self.create_zip_package()
+
+    def clean(self):
+        """Deletes build files and directories."""
+
+        shutil.rmtree(self.build_dir, ignore_errors=True)
+
+    def obtain_server_resources(self):
+        """Builds <server/>."""
+
+        # Copy from <fusion/portableglobe/servers/>:
+        servers_dir = os.path.join(
+            self.source_dir, 'fusion', 'portableglobe', 'servers')
+
+        exclude_entries = ['linux', 'mac', 'windows', 'fileunpacker']
+        entries = [f
+                   for f in os.listdir(servers_dir)
+                   if f.lower() not in exclude_entries]
+        copy_from_dir_to_dir(servers_dir, self.server_dir, entries)
+        copy_from_dir_to_dir(
+            os.path.join(servers_dir, self.platform), self.server_dir)
+
+        # Copy from <maps/>:
+        copy_from_dir_to_dir(
+            os.path.join(self.source_dir, 'maps'),
+            os.path.join(self.server_dir, 'local', 'maps'),
+            exclude_entries=['SConscript'])
+
+        self.create_version_txt()
+
+        # Copy configuration files:
+        copy_from_dir_to_dir(self.resources_dir, self.server_dir)
+
+    def create_version_txt(self):
+        """Creates <server/local/version.txt>."""
+
+        self.get_version()
```

Comment:
`get_version()` only updates values, if they're `None`, so it can be called multiple times without harm.  There really isn't a use case for calling it again right now, but in an imaginary world where someone did something like:

```Python
builder = Builder(build_dir, source_dir, platform)
builder.build_date = datetime.date(1945, 5, 8) # Travel back in time.
builder.version_string = None # Recalculate version_string
builder.build() # Make a build for Doc Brown's trip.
```

It would build with a valid <version.txt>.